lib/gen_stage.ex

```diff
@@ -1722,22 +1757,58 @@ defmodule GenStage do
   def init({mod, args}) do
     case mod.init(args) do
       {:producer, state} ->
-        init_producer(mod, [], state)
+        init_producer(mod, [], state, nil)
+
+      {:producer, state, {:continue, _term} = continue} ->
+        init_producer(mod, [], state, continue)
+
+      {:producer, state, :hibernate} ->
```

Comment:
I would force these return values to be `{:producer, state, [], {:continue, term} | :hibernate}` to simplify the API. Thoughts? Same for the cases below with producer_consumer and consumer.

Reply:
I agree this incredibly long case statement is less than ideal but if I understand your suggestion correctly wouldn't this result in the GenStage not being initialized (validation/state setup) properly at all in the `:hibernate` case and in the `{:continue, term}` case initialization of the GenStage state would be forced upon the user...

Reply:
If you're OK with keeping the current logic as it is it might be cleaner to pull all this out into a pattern matched group of functions instead?

Reply:
Right now you have a few extra cases because of the fact that you are allowing the third element of the tuple to be either the options, `:hibernate`, or `{:continue, term}`. What I am suggesting is that you don't allow it to be the options. The spec of the return would become something like `{:producer, state :: term, options :: keyword, :hibernate | {:continue, term}}`.

This way, you shave off a few case clauses by forcing users to use an empty list of options (`[]`) if they have no options to pass but still want to use `:hibernate` or `{:continue, term}`. Does it make more sense?

Reply:
Right I see, thanks for the clarification! I'll look into these changes tonight 👍